File: lib/test/t_IndependentCopula_std.cxx
//                                               -*- C++ -*-
/**
 *  @file  t_IndependentCopula_std.cxx
 *  @brief The test file of class IndependentCopula for standard methods
 *
 *  Copyright (C) 2005-2013 EDF-EADS-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 *  @author schueller
 *  @date   2012-07-16 15:59:45 +0200 (Mon, 16 Jul 2012)
 */
#include "OT.hxx"
#include "OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

class TestObject : public IndependentCopula
{
public:
  explicit TestObject() : IndependentCopula(3) {}
  explicit TestObject(const String & name) : IndependentCopula(3)
  {
    setName(name);
  }
  virtual ~TestObject() {}
};


int main(int argc, char *argv[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);
  setRandomGenerator();

  try
    {
      // Test basic functionnalities
      //    checkClassWithClassName<TestObject>();

      // Test some extra functionnalities
      //    checkNameFeature<TestObject>();

      // Instanciate one distribution object
      UnsignedLong dim(3);
      IndependentCopula copula(dim);
      fullprint << "Copula " << copula << std::endl;
      std::cout << "Copula " << copula << std::endl;

      // Is this copula elliptical ?
      fullprint << "Elliptical distribution= " << (copula.isElliptical() ? "true" : "false") << std::endl;

      // Is this copula continuous ?
      fullprint << "Continuous = " << (copula.isContinuous() ? "true" : "false") << std::endl;

      // Is this copula elliptical ?
      fullprint << "Elliptical = " << (copula.hasEllipticalCopula() ? "true" : "false") << std::endl;

      // Is this copula independant ?
      fullprint << "Independent = " << (copula.hasIndependentCopula() ? "true" : "false") << std::endl;

      // Test for realization of copula
      NumericalPoint oneRealization = copula.getRealization();
      fullprint << "oneRealization=" << oneRealization << std::endl;

      // Test for sampling
      UnsignedLong size = 10000;
      NumericalSample oneSample = copula.getSample( size );
      fullprint << "oneSample first=" << oneSample[0] << " last=" << oneSample[size - 1] << std::endl;
      fullprint << "mean=" << oneSample.computeMean() << std::endl;
      fullprint << "covariance=" << oneSample.computeCovariance() << std::endl;

      // Define a point
      NumericalPoint point( copula.getDimension(), 0.6 );
      fullprint << "Point= " << point << std::endl;

      // Show DDF, PDF and CDF of point
      NumericalPoint  pointDDF = copula.computeDDF( point );
      NumericalScalar pointPDF = copula.computePDF( point );
      NumericalScalar pointCDF = copula.computeCDF( point );
      fullprint << "point= " << point
                << " ddf=" << pointDDF
                << " ddf (FD)=" << copula.ContinuousDistribution::computeDDF(point)
                << " pdf=" << pointPDF
                << " cdf=" << pointCDF
                << std::endl;
      NumericalPoint PDFgr = copula.computePDFGradient( point );
      fullprint << "pdf gradient     =" << PDFgr << std::endl;
      NumericalPoint CDFgr = copula.computeCDFGradient( point );
      fullprint << "cdf gradient     =" << CDFgr << std::endl;
      NumericalPoint quantile = copula.computeQuantile( 0.95 );
      fullprint << "quantile=" << quantile << std::endl;
      fullprint << "cdf(quantile)=" << copula.computeCDF(quantile) << std::endl;
      NumericalPoint mean = copula.getMean();
      fullprint << "mean=" << mean << std::endl;
      CovarianceMatrix covariance = copula.getCovariance();
      fullprint << "covariance=" << covariance << std::endl;
      IndependentCopula::NumericalPointWithDescriptionCollection parameters = copula.getParametersCollection();
      fullprint << "parameters=" << parameters << std::endl;

      // Specific to this copula

      // Extract the marginals
      for (UnsignedLong i = 0; i < dim; i++)
        {
          Copula margin(copula.getMarginal(i));
          fullprint << "margin=" << margin << std::endl;
          fullprint << "margin PDF=" << margin.computePDF(NumericalPoint(1, 0.25)) << std::endl;
          fullprint << "margin CDF=" << margin.computeCDF(NumericalPoint(1, 0.25)) << std::endl;
          fullprint << "margin quantile=" << margin.computeQuantile(0.95) << std::endl;
          fullprint << "margin realization=" << margin.getRealization() << std::endl;
        }

      // Extract a 2-D marginal
      Indices indices(2, 0);
      indices[0] = 1;
      indices[1] = 0;
      fullprint << "indices=" << indices << std::endl;
      Copula margins(copula.getMarginal(indices));
      fullprint << "margins=" << margins << std::endl;
      fullprint << "margins PDF=" << margins.computePDF(NumericalPoint(2, 0.25)) << std::endl;
      fullprint << "margins CDF=" << margins.computeCDF(NumericalPoint(2, 0.25)) << std::endl;
      quantile = margins.computeQuantile(0.95);
      fullprint << "margins quantile=" << quantile << std::endl;
      fullprint << "margins CDF(quantile)=" << margins.computeCDF(quantile) << std::endl;
      fullprint << "margins realization=" << margins.getRealization() << std::endl;


    }
  catch (TestFailed & ex)
    {
      std::cerr << ex << std::endl;
      return ExitCode::Error;
    }


  return ExitCode::Success;
}

File: python/test/t_IndependentCopula_std.py
#! /usr/bin/env python

from openturns import *

TESTPREAMBLE()
RandomGenerator().SetSeed(0)

try :
    # Instanciate one distribution object
    dim = 3

    copula = IndependentCopula(dim)
    print "Copula =", repr(copula)
    print "Copula =", copula

    # Is this copula an elliptical distribution?
    print "Elliptical distribution = ", copula.isElliptical()

    # Is this copula continuous ?
    print "Continuous copula = ",  copula.isContinuous()

    # Is this copula elliptical ?
    print "Elliptical copula = ", copula.hasEllipticalCopula()

    # Is this copula independant ?
    print "hasIndependentCopula = ",copula.hasIndependentCopula()

    # Test for realization of copula
    oneRealization = copula.getRealization()
    print "oneRealization=", repr(oneRealization)

    # Test for sampling
    size = 10000
    oneSample = copula.getSample( size )
    print "oneSample first=" , repr(oneSample[0]) , " last=" , repr(oneSample[size - 1])
    print "mean=" , repr(oneSample.computeMean())
    print "covariance=" , repr(oneSample.computeCovariance())

    # Define a point
    point = NumericalPoint(copula.getDimension(), 0.6)
    print "Point= " , repr(point)

    # Show PDF and CDF of point

    # derivative of PDF with regards its arguments
    DDF = copula.computeDDF( point )
    print "ddf     =" , repr(DDF)

    # PDF value
    PDF = copula.computePDF( point )
    print "pdf     =%.6f" % PDF

    # CDF value
    CDF = copula.computeCDF( point )
    print "cdf=%.6f" % CDF

    # derivative of the PDF with regards the parameters of the distribution
    PDFgr = copula.computePDFGradient( point )
    print "pdf gradient     =" , repr(PDFgr)

    # derivative of the CDF with regards the parameters of the distribution
    CDFgr = copula.computeCDFGradient( point )
    print "cdf gradient     =" , repr(CDFgr)

    # quantile
    quantile = copula.computeQuantile( 0.95 )
    print "quantile=" , repr(quantile)
    print "cdf(quantile)=%.6f" % copula.computeCDF(quantile)

    # mean
    mean = copula.getMean()
    print "mean=" , repr(mean)

    # covariance
    covariance = copula.getCovariance()
    print "covariance=" , repr(covariance)

    # parameters of the distribution
    parameters = copula.getParametersCollection()
    print "parameters=" , repr(parameters)

    # Specific to this copula

    # Extract the marginals
    for i in range(dim) :
        margin = copula.getMarginal(i)
        print "margin=", repr(margin)
        print "margin PDF=%.6f" % margin.computePDF(NumericalPoint(1,0.25))
        print "margin CDF=%.6f" % margin.computeCDF(NumericalPoint(1,0.25))
        print "margin quantile=", repr(margin.computeQuantile(0.95))
        print "margin realization=", repr(margin.getRealization())

    # Extract a 2-D marginal
    indices = Indices(2, 0)
    indices[0] = 1
    indices[1] = 0
    print "indices=", repr(indices)
    margins = copula.getMarginal(indices)
    print "margins=", repr(margins)
    print "margins PDF=%.6f" % margins.computePDF(NumericalPoint(2,0.25))
    print "margins CDF=%.6f" % margins.computeCDF(NumericalPoint(2,0.25))
    quantile = margins.computeQuantile(0.95)
    print "margins quantile=", repr(quantile)
    print "margins CDF(quantile)=%.6f" % margins.computeCDF(quantile)
    print "margins realization=", repr(margins.getRealization())

except :
    import sys
    print "t_IndependantCopula.py", sys.exc_type, sys.exc_value
